### PR TITLE
NF: Rudimentary web-face for datasets.datalad.org

### PR DESCRIPTION
**Pull request by debanjum**
currently datalad ls -w results in filesystem traversal at
mentioned location. stores data in list with a flat
sublist of files, directories per git/annex submodule.
can handle recursive case too.

Next: create model, that can by passed to the json formatter
ToDo:  create json formatter, create web UI renderer based on json
This is for the NF #520

### CODE REVIEW

File: datalad/interface/ls.py [hunk 1]
```diff
@@ -189,6 +196,89 @@ def annex_local_size(self):
...
+            return 0
+        elif 'link' in _type:
+            return lstat(self._symlink).st_size
+        elif 'dir' in _type:
+            return lstat(self._path).st_size  # add du -s command for plain dir
+        else:
+            return -1
+
+    @property
+    def _type(self):
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: Convention is to append _ as a suffix, not as a prefix if just need to avoid collision with the built in defs

File: datalad/interface/ls.py [hunk 2]
```diff
@@ -189,6 +196,89 @@ def annex_local_size(self):
...
+            return None
+
+    @property
+    def date(self):
+        """Date of last modification
+        """
+        if self._type is not ['git', 'annex']:
+            return lstat(self._path).st_mtime
+        else:
+            super(self.__class__, self).date
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: Cute ;-) but ideally we should check the date of the last commit which has changed that file... should be relatively fast if done via GitPython

File: datalad/interface/ls.py [hunk 3]
```diff
@@ -189,6 +196,89 @@ def annex_local_size(self):
...
+
+    @property
+    def size(self):
+        _type = self._type
+        if not _type:
+            return -1
+        if 'annex' in _type:
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: Hm.. why in and not just ==?

File: datalad/interface/ls.py [hunk 4]
```diff
@@ -189,6 +196,89 @@ def annex_local_size(self):
...
+            return self.annex_local_size
+        elif 'git' in _type:
+            return self.git_local_size
+        elif 'file' in _type:
+            return lstat(self._path).st_size
+        elif 'broken-link' in _type:
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: Ideally we should deduce from git annex key - it has size encoded there... git annex info --fast could be used as well - there is AnnexRepo.info(..., batch=True) to help... you better reuse the instance of repo across files of the same repository to take advantage of batch execution

File: datalad/interface/ls.py [hunk 5]
```diff
@@ -189,6 +196,89 @@ def annex_local_size(self):
...
+        elif 'link' in _type:
+            return lstat(self._symlink).st_size
+        elif 'dir' in _type:
+            return lstat(self._path).st_size  # add du -s command for plain dir
+        else:
+            return -1
+
+    @property
+    def _type(self):
+        if islink(self.path):
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: Here we should also better ask AnnexRepo either file has content or not... we might publish to some crippled filesystem which would get annex into direct mode, so no symlinks would be there

File: datalad/interface/ls.py [hunk 6]
```diff
@@ -189,6 +196,89 @@ def annex_local_size(self):
...
+            return 'git'
+        elif isdir(self.path):
+            return 'dir'
+        else:
+            return None
+
+    @property
+    def git_local_size(self):
+        try:
+            describe, outerr = self.repo._git_custom_command([], ['git', 'count-objects', '-v'])[0].split('\n')
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: Yet to figure out what this one does... having a docstring above would have helped ;-)

File: datalad/interface/ls.py [hunk 7]
```diff
@@ -297,6 +386,75 @@ def _ls_dataset(loc, fast=False, recursive=False, all=False):
         ds_str = format_ds_model(formatter, dsm, full_fmt, format_exc=path_fmt + u"  {msg!R}")
         print(ds_str)
 
+
+def JsonFormatter(path, repo, _type, size, date):
+    pretty_size = humanize.naturalsize(size)
+    pretty_date = time.strftime(u"%Y-%m-%d/%H:%M:%S", time.localtime(date))
+    json_fmt = u'{\"path\": \"%s\", \"repo\": \"%s\", \"type\": \"%s\", \"size\": \"%s\", \"date\": \"%s\"}'
+    return json_fmt % (path, repo, _type, pretty_size, pretty_date)
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: I don't think you should bother with manual json formatting... create a dict with all fields, use json.dumps (iirc function correctly). git grep json  since we dump json files in a few places

File: datalad/interface/ls.py (hunk 4 above)
Comment (debanjum) [inline datalad/interface/ls.py]: You mean associate a single instance of the parent repo class (i.e GitRepo or AnnexRepo) with each file in a given repository ?

Comment (yarikoptic) [inline datalad/interface/ls.py]: Yes

File: datalad/interface/ls.py (hunk 6 above)
Comment (debanjum) [inline datalad/interface/ls.py]: it computes the disk space used by unpacked object files in the git repository.
thought maybe more appropriate to use for a git directory instead of `os.lstat` or `du -s`

Comment (yarikoptic) [inline datalad/interface/ls.py]: Actually I like the idea now, if it is not taking too long to run... so we could report how much git takes, how much of annex load within any particular dataset or collection of them... would be useful to guesstimate amount of traffic necessary for cloning..  we might even use this information later during install

File: datalad/resources/website/index.html [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		     { "data": "date", title: "Date" },
+		     { "data": "path", title: "Path" },
+		     { "data": "type", title: "Type" },
+		 ],
+		 //update css of row based on type of node
+		 "createdRow": function ( row, data, index ) {
+		     if ( data.type == 'dir' ) {
+			 $('td', row).eq(0).addClass('dir');
+			 $('td', row).eq(1).addClass('dir');
+			 $('td', row).eq(2).addClass('dir');
```
Comment (yarikoptic) [inline datalad/resources/website/index.html]: Duplication is evil ;-)

**Comment by yarikoptic** [comment]
Any live example to see?

File: datalad/resources/website/index.html (hunk 8 above)
Comment (debanjum) [inline datalad/resources/website/index.html]: Yeah, i know! :grin: 
I'm going to clean all this up soon.

**Comment by yarikoptic** [comment]
so the test listing is available at  http://datasets.datalad.org/webtest/

**Comment by yarikoptic** [comment]
please merge current master

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 9]
```diff
@@ -284,13 +285,28 @@ def __call__(sshurl, target=None, target_dir=None,
...
-                     opj(path, ".git/hooks/post-update")])
+                json_command = '"datalad ls -r --json file ' + path + '"'
+                virtualenv = 'source /home/debanjum/datalad/.env/bin/activate'  # NOT_SCALABLE! custom virtualenv path
+                for post_update_hook in ['"#!/bin/bash"', '"git update-server-info"', virtualenv, json_command]:
+                    ssh(["echo", post_update_hook, '>>', opj(path, '.git/hooks/post-update')])
+                ssh(["chmod", '+x', opj(path, '.git/hooks/post-update')])  # make the script executable
             except CommandError as e:
-                lgr.error("Failed to enable post update hook.\n"
+                lgr.error("Failed to add json creation command to post update hook.\n"
+                          "Error: %s" % e.message)
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: Generally better to use `exc_str` helper while logging information about caught exception, since then we could control of how much information (traceback) is provided within the message -- came really handy few times

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 10]
```diff
@@ -284,13 +285,28 @@ def __call__(sshurl, target=None, target_dir=None,
...
+
+            # get html for dataset rendering from local datalad repo:
+            lgr.info("Uploading web interface ...")
+            try:
+                import datalad
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: why import not at the top of the file?

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 11]
```diff
@@ -284,13 +285,28 @@ def __call__(sshurl, target=None, target_dir=None,
...
+                html = opj(dirname(datalad.__file__), "resources/website/index.html")
+                scp_target = sshri.hostname + ":" + sshri.path
+                Runner().run(['scp', '-P', sshri.port, html, scp_target] if sshri.port else ['scp', html, scp_target])
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: ideally we should avoid direct scp calls like this.  There is this `ssh` manager which should be our central point of interaction with ssh.  Now you would not take advantage of controlpath and not sure how `sshri.port` would play out here if not specified in the url.  So ideally I think we should provide  `SSHConnection.copy(path, remote_dest, ...)` to provide 'scp' operation, may be with support for options such as recursion, preserving permissions etc... 
Then it should also be used above instead of the chain of "echo" calls to create post-update hook.  You can generate entire temp file locally and then .copy it there.

File: datalad/interface/ls.py [hunk 12]
```diff
@@ -189,6 +201,88 @@ def annex_local_size(self):
...
+
+    def __init__(self, path, repo=None):
+        self._path = path  # fs path to the node, can be overridden
+        self._info = None
+        self.repo = DsModel(Dataset(path)) if self.type_ == 'annex' else repo  # parent repository associated with node
+        self._branch = None
+
+    @property
+    def symlink(self):
+        """if symlink returns path the symlink points to else returns False"""
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: IMHO better to return `None`  (yet to see how it is used).  It would eval to False if asked as 'bool', but also would have clearer semantic I think

File: datalad/interface/ls.py [hunk 13]
```diff
@@ -193,6 +205,88 @@ def annex_local_size(self):
...
+        elif 'git' == type_:
+            return self.git_local_size
+        elif 'file' == type_:
+            return lstat(self._path).st_size
+        elif type_ in ['link', 'link-broken']:
+            if self.repo.repo.is_under_annex(self._path):
+                return self.repo.repo.info(self._path, batch=True)['size']
+            return 0 if type_ == 'broken-link' else lstat(self.symlink).st_size
+        elif 'dir' == type_:
+            return Runner().run(['du', '-s', self._path])[0].split('\t')[0]  # lstat(self._path).st_size
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: to be gone!

File: datalad/interface/ls.py [hunk 14]
```diff
@@ -193,6 +205,88 @@ def annex_local_size(self):
...
+    def size(self):
+        """Size of the node computed based on its type"""
+        type_ = self.type_
+        if not type_:
+            return -1
+        if 'annex' == type_:
+            if self.annex_local_size:
+                return self.annex_local_size
+            else:
+                return Runner().run(['du', '-s', self._path])[0].split('\t')[0]
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: to be gone

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 15]
```diff
@@ -284,13 +285,35 @@ def __call__(sshurl, target=None, target_dir=None,
...
+                with open(tempf, 'a') as f:  # create post_update hook script in local tempfile
+                    for post_update_cmd in ['#!/bin/bash', 'git update-server-info', virtualenv, json_command]:
+                        f.write(post_update_cmd + '\n')
+                # upload hook to dataset
+                Runner().run(['scp', '-P', sshri.port, tempf, hook_ssh] if sshri.port else ['scp', tempf, hook_ssh])
+                ssh(["chmod", '+x', opj(path, '.git/hooks/post-update')])  # make it executable
             except CommandError as e:
-                lgr.error("Failed to enable post update hook.\n"
+                lgr.error("Failed to add json creation command to post update hook.\n"
+                          "Error: %s" % e.message)
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: use `exc_str(e)` not `e.message`

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 16]
```diff
@@ -284,13 +285,35 @@ def __call__(sshurl, target=None, target_dir=None,
...
+                tempf = Runner().run(['mktemp'])[0].split('\n')[0]
+                hook_loc = opj(path, '.git/hooks/post-update')
+                hook_ssh = sshri.hostname + ":" + hook_loc
+
+                with open(tempf, 'a') as f:  # create post_update hook script in local tempfile
+                    for post_update_cmd in ['#!/bin/bash', 'git update-server-info', virtualenv, json_command]:
+                        f.write(post_update_cmd + '\n')
+                # upload hook to dataset
+                Runner().run(['scp', '-P', sshri.port, tempf, hook_ssh] if sshri.port else ['scp', tempf, hook_ssh])
+                ssh(["chmod", '+x', opj(path, '.git/hooks/post-update')])  # make it executable
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: this will be refactored after scp #633 is done

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 17]
```diff
@@ -284,13 +285,35 @@ def __call__(sshurl, target=None, target_dir=None,
...
+            lgr.info("Uploading web interface ...")
+            try:
+                import datalad
+                html = opj(dirname(datalad.__file__), "resources/website/index.html")
+                scp_target = sshri.hostname + ":" + sshri.path
+                Runner().run(['scp', '-P', sshri.port, html, scp_target] if sshri.port else ['scp', html, scp_target])
+
+            except CommandError as e:
+                lgr.error("Failed to get html from local datalad repository. Unable to setup web interface.\n"
                           "Error: %s" % e.message)
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: `exc_str`

File: datalad/interface/ls.py [hunk 18]
```diff
@@ -193,6 +205,88 @@ def annex_local_size(self):
...
+            return None
+
+    @property
+    def git_local_size(self):
+        """computes the disk space used by unpacked object files in the git repository"""
+        try:
+            describe, outerr = self.repo._git_custom_command([], ['git', 'count-objects', '-v'])[0].split('\n')
+            size = [item for item in describe if 'size: ' in item][0].split(': ')
+            return int(size[1])
+        except:
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: catch actual exception and use `exc_str`
but first check why it fails at all, try to make it conditioned without invoking git without necessity

File: datalad/interface/ls.py [hunk 19]
```diff
@@ -193,6 +205,88 @@ def annex_local_size(self):
...
+        elif isdir(self.path):
+            return 'dir'
+        else:
+            return None
+
+    @property
+    def git_local_size(self):
+        """computes the disk space used by unpacked object files in the git repository"""
+        try:
+            describe, outerr = self.repo._git_custom_command([], ['git', 'count-objects', '-v'])[0].split('\n')
```
Comment (yarikoptic) [inline datalad/interface/ls.py]: move into GitRepo class as a separate PR `get_object_stats`, with a proper'ish test

**Comment by yarikoptic** [comment]
Please push updated view to http://datasets.datalad.org/webtest/

**Comment by yarikoptic** [comment]
@debanjum 
so how fast/slow it is on those sample datasets? ;)  should we try running it on the entirety?
could you please update http://datasets.datalad.org/webtest/ so it doesnt blow?

**Comment by debanjum** [comment]
I've published ds000110 to http://datasets.datalad.org/webtest/ for now.

Attempts at running on the complete openfmri failed at `create-publication` complaining about no path to the `openfmri/ds000171` dataset
```bash
source ~/***/datalad/venv/bin/activate
datalad install --recursive yes http://datasets.datalad.org/openfmri && cd openfmri
SERVER_URL="ssh://***/webtest"
datalad create-publication-target-sshwebserver --existing=replace --recursive $SERVER_URL
```
```sh
2016-08-09 18:28:41,552 [ERROR  ] <class 'git.exc.NoSuchPathError'>: /***/openfmri/ds000171 (gitrepo.py:479)
2016-08-09 18:28:41,580 [ERROR  ] /***/openfmri/ds000171 [base.py:__init__:139] (NoSuchPathError) (main.py:246)
```

**Comment by yarikoptic** [comment]
thanks! fixed up ds000171 ... not sure how that has happened since it better doesn't ;)  try again
but in the long(er) run we might to add similar to 'publish' command option to `--skip-failing`

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 20]
```diff
@@ -295,21 +299,39 @@ def __call__(sshurl, target=None, target_dir=None,
                             "You will not be able to push to checked out "
                             "branch. Error: %s", path, exc_str(e))
 
-            # enable post-update hook:
+            # create post_update hook to refresh dataset metadata on publication server
+            lgr.info("Enabling git post-update hook ...")
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: hm, how come this code is claimed to be not tested... I thought there is at least a smoke test using ssh.  please check/fix

Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: ideally there should be a unittest for a complete loop where dataset is getting published and then verified that publication went correctly

File: datalad/interface/tests/test_ls.py [hunk 21]
```diff
@@ -58,6 +64,162 @@ def test_ls_repos(toppath):
...
+    meta_path = opj(topdir, meta_dir)
+
+    for all in [True, False]:
+        for recursive in [True, False]:
+            for state in ['file', 'delete']:
+                with swallow_logs(), swallow_outputs():
+                    ds = _ls_json(topdir, json=state, all=all, recursive=recursive)
+
+                # subdataset should have its json created and deleted when all=True else not
+                subds_metahash = hashlib.md5('/'.encode('utf-8')).hexdigest()
```
Comment (yarikoptic) [inline datalad/interface/tests/test_ls.py]: duplication is evil! ;)  make a function `get_metahash` or smth like that, which would get your path, return a hash.  Use here and elsewhere

File: datalad/interface/tests/test_ls.py [hunk 22]
```diff
@@ -58,6 +64,162 @@ def test_ls_repos(toppath):
...
+    git = GitRepo(opj(topdir, 'dir', 'subgit'), create=True)                    # create git repo
+    git.add(opj(topdir, 'dir', 'subgit', 'fgit.txt'), commit=True)              # commit to git to init git repo
+    annex.add(opj(topdir, 'dir', 'subgit'), commit=True)                        # add the non-dataset git repo to annex
+    annex.add(opj(topdir, 'dir'), commit=True)                                  # add to annex (links)
+    annex.drop(opj(topdir, 'dir', 'subdir', 'file2.txt'), options=['--force'])  # broken-link
+
+    meta_dir = '.git/datalad/metadata'
+    meta_path = opj(topdir, meta_dir)
+
+    for all in [True, False]:
```
Comment (yarikoptic) [inline datalad/interface/tests/test_ls.py]: make `all` `all_` to not overload the builtin

File: datalad/resources/website/index.html [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+	 td.link {color: #FFD03C;  }
+	 td.file {color: #67CDDC;  }
+	</style>
+    </head>
+    <body>
+        <h1><img src='http://datalad.org/pics/favicon.svg' width="24">Datalad Repository</h1>
+	<table id="directory" class="display" width="100%" cellspacing="0"></table>
+	<script type="text/javascript" charset="utf8" src="http://ajax.aspnetcdn.com/ajax/jQuery/jquery-1.12.3.min.js"></script>
+	<script type="text/javascript" charset="utf8" src="http://ajax.aspnetcdn.com/ajax/jquery.dataTables/1.10.5/jquery.dataTables.min.js"></script>
+	<script type="text/javascript" charset="utf8" src="https://raw.githubusercontent.com/blueimp/JavaScript-MD5/master/js/md5.min.js"></script>
```
Comment (yarikoptic) [inline datalad/resources/website/index.html]: we would need a copy of original sources within our git repo somewhere under `datalad/resources/www/js`... eventually we might want to deploy them instead of referring to aspnetcdn, so they could be installed alongside (offline use, not depending on github and aspnetcdn being online etc).  Also extend COPYING then with license on those pieces which you would include.

File: datalad/resources/website/index.html [hunk 24]
```diff
@@ -0,0 +1,123 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="UTF-8">
+        <title>Datalad Repository</title>
+        <link rel="shortcut icon" type="image/x-icon" href="http://datalad.org/pics/favicon-16x16.png" />
+	<link rel="stylesheet" type="text/css" href="http://ajax.aspnetcdn.com/ajax/jquery.dataTables/1.10.5/css/jquery.dataTables.css">
```
Comment (yarikoptic) [inline datalad/resources/website/index.html]: and this one should got into resources

File: datalad/resources/website/index.html [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+		 "paging": false,  //ensure scrolling instead of pages
+		 "ajax": {         //specify url to get json from ajax
+		     "url": metadata_url,
+		     "dataSrc": "nodes"
+		 },
+	         "order": [[ 6, "desc" ], [ 0, 'asc' ]],
+		 "columns": [      //select columns and their names from json
+		     { data: "name", title: "Name" },
+                     { data: "date", title: "Last Modified", className: "dt-center", width: "20%" },
+                     { data: "size", title: "Size", className: "dt-center" },
```
Comment (yarikoptic) [inline datalad/resources/website/index.html]: is there some meaning to varying indentation here?

**Comment by hanke** [comment]
@debanjum @yarikoptic Tests pass! I'd be interested in a quick "here is what you need to do to test this out". Thanks!

**Comment by yarikoptic** [comment]
Go to http://datasets.datalad.org/webtest/

**Comment by debanjum** [comment]
To just see how the UI looks refer http://datasets.datalad.org/webtest/ as @yarikoptic mentioned

To play around with it:
 You can either publish the datasets to a server with `create-publication-target` and `publish`
```bash
cd /a/local/dataset/folder
SERVER_URI="ssh://[your_server_uri]"

# create remote, enable metadata updation with post-update hook, push web-ui(index.html)
datalad create-publication-target-sshwebserver --existing=replace --recursive $SERVER_URI public-server

# publish dataset to server, also triggers the post-update hook 
# the post-update hook runs the create/update metadata json command on server
datalad publish -r --to public-server .

# open the server's url. e.g @ http://datasets.datalad.org/webtest/
```
OR

You can test it locally by manually running the `datalad ls` command
```bash
cd /a/local/dataset/folder
# create the metadata for the dataset
# -a for recursive traversal of subdatasets
# -r for recursive traversal of the current dataset's files (except subdatasets)
datalad ls -a -r --json file .
# get UI to dataset root
cp $DATALAD_FOLDER/datalad/resources/website/index.html ./
# start server for testing
python -m SimpleHTTPServer

# visit [locahost]:[port] to see the UI
```

Published in the interest of breaking things before they're in master
@hanke

**Comment by debanjum** [comment]
The UI has been put together a bit haphazardly, metadata creation takes ~10mins for openfmri, etc.
So don't expect polish but if everything runs then thats at least a start.

File: datalad/resources/website/index.html (hunk 25 above)
Comment (debanjum) [inline datalad/resources/website/index.html]: my emacs shows them w/o the varying indentation.
seems like more of a tabs vs spaces editor rendering issue :/
made it all spaces for now. hopefully that fixes it.

File: datalad/resources/website/index.html (hunk 23 above)
Comment (debanjum) [inline datalad/resources/website/index.html]: Yeah, but should I change them to pull from our github repo for now, i.e somethings like https://raw.githubusercontent.com/datalad/datalad/example.js ? It would be pretty slow that way actually and the way i'm pulling the md5.js script isn't ideal either.
I could put the css, js files under '.git/datalad' of the dataset too on `create-publication`
this is where the generated metadata is currently being stored, don't know if its a good idea though

File: datalad/interface/tests/test_ls.py (hunk 21 above)
Comment (yarikoptic) [inline datalad/interface/tests/test_ls.py]: @debanjum seems still to be TODO

File: datalad/resources/website/index.html (hunk 23 above)
Comment (yarikoptic) [inline datalad/resources/website/index.html]: not sure what is the ideal setup would be.  I see possible scenarios, but in each of those we better have copies of those js/css under our `datalad/resources/web` as I have said before:
1. fully self sufficient "independent" -- all js/css are deployed for every dataset. *pros*: guaranteed to be consistent with index.html script deployed within that dataset, not dependent on github etc being available/online.  *cons*: for every dataset browser would need to fetch exactly the same (often) copy of them, thus slow
2. fully self sufficient "top repo only": we revert back to provide index.html only within top dataset and then revert back and do not have those "jumps" into subdatasets urls, so path to sub-datasets would always be within path option of url (e.g. http://datasets.datalad.org/webtest/?dir=ds000001/).  And then deploy all those .js /.css files for the super-dataset.  *pros*: all of 1. without cons of 1 since we have a single copy/url to get to them. Might also be easier to allow users for customizations *cons*: no "direct" urls to sub-datasets (only through path= option, rewrites on server side probably possible if necessary)
3. dependent on some delivery website (like githubusercontent):  so we have our own versioned copies (under `datalad/resources/web`) and point urls to them.  *pros*: might be faster access to them from github?  *cons* not guaranteed access to them, couldn't be used offline (e.g. from some quick local webserver offline)

So I think 2. is probably the way to go unless I see some showstopper. @hanke -- welcome to chime in.

Comment (yarikoptic) [inline datalad/resources/website/index.html]: BTW, anyways for `create-publication-target-sshwebserver` we need to add `--existing=reconfigure` or even more specific `--existing=reconfigure-web` which would not replace entire existing repo, but just reconfigures for the web (regenerates the hook, deposits current versions of index/.js/.css).  So please add that mode within this PR @debanjum

**Comment by debanjum** [comment]
Added js testing to project.

Basic notes to setup env and run tests 
```sh
npm install grunt-cli
npm install grunt grunt-contrib-qunit
grunt test -verbose
```
[qunit](https://qunitjs.com/) is the unit-testing framework being used
[sinon.js](http://sinonjs.org/qunit/) qunit plugin is used for mocking functions and servers request-responses
[grunt](http://gruntjs.com/) for the test automation
travis uses [phantomjs](http://phantomjs.org/)-a headless browser for the front-end testing by default

**Comment by debanjum** [comment]
Fixed up a [few](https://github.com/datalad/datalad/pull/541/commits/99fc7a9d7b7d78acee1ca2d9738a5616026870fe) [issues](https://github.com/datalad/datalad/pull/541/commits/bf94ff2067dc6b647bc947728fa4de737f314cfc), the final results are on datasets.datalad.org/webtest.
I think this branch is ready for a final review

**Comment by yarikoptic** [comment]
obviously tests are failing due to changes in this branch... but couldn't yet reproduce locally. and in precise schroot it just stalls for me 

```shell
*$>  DATALAD_TESTS_SSH=1 nosetests -s -v datalad/distribution/tests/test_target_ssh.py 
datalad.distribution.tests.test_target_ssh.test_target_ssh_recursive ... ok
datalad.distribution.tests.test_target_ssh.test_target_ssh_simple ... 
```
with loglevel debug it is here: (I guess annex is trying to ask smth)
```
2016-08-27 17:12:07,068 [DEBUG  ] Running: ['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'annex', 'enableremote', '--debug', 'local_target'] (cmd.py:352)
```
so, I guess the easiest way would be to just backdoor travis instance (unless they finally came up with official way to debugging sessions -- please check).

also note that test_runner failed now whenever there is a space in symlinked tmpdir:
https://travis-ci.org/datalad/datalad/jobs/155515465
which didn't happen before

also please close (click x) on all those numerous unneeded intermediate reports from coverages

**Comment by yarikoptic** [comment]
FYI you've reached 100 commits boundary after which  buildbot doesn't run tests any longer :-/

**Comment by yarikoptic** [comment]
replaced with #738
